lib/protein/adapters/amqp_adapter/server.ex

```diff
@@ -14,43 +14,66 @@ defmodule Protein.AMQPAdapter.Server do
   def init(opts) do
     Process.flag(:trap_exit, true)
     chan = connect(opts)
-    {:ok, {chan, opts, 0}}
+    state = %{channel: chan, opts: opts, consumers: []}
+    {:ok, state}
   end
 
-  def terminate(_reason, {_chan, _opts, running_processes_count}) do
-    wait_for_all_processes(running_processes_count)
+  def terminate(_reason, %{consumers: consumers}) do
+    wait_for_all_consumers(consumers)
   end
 
-  def wait_for_all_processes(0), do: :ok
+  def wait_for_all_consumers([]), do: :ok
 
-  def wait_for_all_processes(running_processes_count) do
+  def wait_for_all_consumers(consumers) do
     receive do
-      {:DOWN, _, :process, _, :normal} -> wait_for_all_processes(running_processes_count - 1)
-      _ -> wait_for_all_processes(running_processes_count)
+      {:DOWN, _, :process, down_pid, :normal} ->
+        present_consumers = Enum.filter(consumers, fn %{pid: pid} -> pid != down_pid end)
+        wait_for_all_consumers(present_consumers)
+
+      _ ->
+        wait_for_all_consumers(consumers)
     end
   end
 
   def handle_info({:basic_consume_ok, _meta}, state), do: {:noreply, state}
   def handle_info({:basic_cancel, _meta}, state), do: {:stop, :normal, state}
   def handle_info({:basic_cancel_ok, _meta}, state), do: {:noreply, state}
 
-  def handle_info({:basic_deliver, payload, meta}, {chan, opts, processes}) do
-    pid = spawn(fn -> consume(chan, opts, meta, payload) end)
-    Process.monitor(pid)
-    {:noreply, {chan, opts, processes + 1}}
+  def handle_info(
+        {:basic_deliver, payload, meta},
+        %{
+          channel: chan,
+          opts: opts,
+          consumers: consumers
+        } = state
+      ) do
+    {pid, ref} = spawn_monitor(fn -> consume(chan, opts, meta, payload) end)
+    {:noreply, %{state | consumers: consumers ++ [%{pid: pid, meta: meta, monitor_ref: ref}]}}
   end
 
   def handle_info(
         {:DOWN, _, :process, pid, _reason},
-        {%Channel{conn: %Connection{pid: pid}}, opts, _}
+        %{channel: %Channel{conn: %Connection{pid: pid}}, consumers: consumers, opts: opts} = state
       ) do
+    kill_consumers(consumers)
     chan = connect(opts)
-    {:noreply, {chan, opts, 0}}
+    {:noreply, %{state | channel: chan, consumers: []}}
+  end
+
+  # handles consumer normal exit
+  def handle_info({:DOWN, _, :process, down_pid, :normal}, %{consumers: consumers} = state) do
+    {_consumer, remaining_consumers} = get_consumer(consumers, down_pid)
+    {:noreply, %{state | consumers: remaining_consumers}}
   end
 
-  # handles normal exit and error
-  def handle_info({:DOWN, _, :process, _pid, _reason}, {chan, opts, count}) do
-    {:noreply, {chan, opts, count - 1}}
+  # handles consumer error
+  def handle_info(
```

Comment:
I don't get how we're going to tell apart these "normal exits" and "consumer errors". Latter ones are supposed to be generated by `spawn_monitor`, but `ref` is not being matched here in any way.